### PR TITLE
feat: new did integration

### PR DESCRIPTION
**Pull request by rflechtner**
#### fixes KILTProtocol/ticket#1504

Design document / examples: https://hackmd.io/94gscuIvRW-8Uw9fpgik9Q?view

This PR fully integrates DIDs as the canonical identity provider in the KILT SDK for credentials, messaging and delegations.
With the proposed changes, authentication and authorisation throughout these systems rely on a DID and associated metadata which bundles a set of keys under an identifier (=DID).
To do so, several new components and interfaces are introduced:
- The DidDetails interface and class, which represent the data associated with a DID, namely a set of public keys, their verification relationships (=purposes, e.g. `CapabilityDelegation` for a key that is used to authorise delegations), plus additional service endpoints.
- The Keystore interface (and an unsafe demo implementation). This component would be the central storage for private key material and provides signing and encryption capabilities to the rest of the system.
- A DidResolver interface and default implementation, which resolves a DID to DidDetails by querying the blockchain.

These changes are linked to two companion PRs on the mashnet node and types repository, which are required for functional integration with the chain:
- https://github.com/KILTprotocol/mashnet-node/pull/239
- https://github.com/KILTprotocol/type-definitions/pull/16

Note that the actors package has been removed here. The plan is to re-add it as part of a new PR which refactors it to work with the new identity concept. I opened a PR as a reminder for this:
- https://github.com/KILTprotocol/sdk-js/pull/407.

## How to test:
I fixed unit and integration tests, but I advise to test and try out thoroughly.
Remember to use the types and node versions referenced above (yalc & building the rust code may be required)

## Checklist:

- [ ] I have merged, (pre-)published and linked the changes on https://github.com/KILTprotocol/type-definitions/pull/16
- [ ] I have merged and (pre-)published the changes on https://github.com/KILTprotocol/mashnet-node/pull/239
- [x] I have verified that the code works
- [ ] I have verified that the code is easy to understand
  - [ ] If not, I have left a well-balanced amount of inline comments
- [ ] I have [left the code in a better state](https://deviq.com/principles/boy-scout-rule)
- [ ] I have documented the changes (where applicable)
  - [ ] Updated example.ts
  - [ ] Updated getting-started{.ts, .md}
  - [ ] Updated JSDoc examples
  - [ ] Updated documentation code examples / api documentation

### CODE REVIEW

File: packages/chain-helpers/package.json [hunk 1]
```diff
@@ -22,16 +22,16 @@
   "homepage": "https://github.com/KILTprotocol/sdk-js#readme",
   "devDependencies": {
     "rimraf": "^3.0.2",
-    "typescript": "4.0"
+    "typescript": "^4.2.2"
   },
   "dependencies": {
     "@kiltprotocol/config": "workspace:*",
-    "@kiltprotocol/type-definitions": "0.1.10",
+    "@kiltprotocol/type-definitions": "file:.yalc/@kiltprotocol/type-definitions",
```
Comment (rflechtner) [inline packages/chain-helpers/package.json:29]: release required (https://github.com/KILTprotocol/type-definitions/pull/16)

File: packages/core/src/__integrationtests__/utils.ts [hunk 2]
```diff
@@ -1,32 +1,36 @@
...
+import { KeyringPair } from '@polkadot/keyring/types'
+import { randomAsU8a } from '@polkadot/util-crypto'
 import BN from 'bn.js/'
 import CType from '../ctype/CType'
 import { getOwner } from '../ctype/CType.chain'
-import Identity from '../identity/Identity'
 
-export const MIN_TRANSACTION = new BN(100_000_000)
-export const ENDOWMENT = MIN_TRANSACTION.mul(new BN(100))
+export const EXISTENTIAL_DEPOSIT = new BN(10 ** 13)
```
[the remark this answers is not in the document]
Comment (rflechtner) [reply packages/core/src/__integrationtests__/utils.ts:10]: Wait, didn't we look this up? 
image

File: packages/kilt-did/src/types.ts [hunk 3]
```diff
@@ -173,58 +149,37 @@ export interface DidPublicKeyDetails extends Struct {
...
-export interface DidUpdateOperation extends Struct {
-  did: DidIdentifier
-  newAuthenticationKey: Option<DidVerificationKey>
-  newKeyAgreementKeys: BTreeSet<DidEncryptionKey>
-  attestationKeyUpdate: DidKeyUpdateAction
-  delegationKeyUpdate: DidKeyUpdateAction
-  publicKeysToRemove: BTreeSet<KeyId>
-  newEndpointUrl: Option<Url>
-  txCounter: u64
+export interface DidRecord extends Struct {
```
[the remark this answers is not in the document]
Comment (rflechtner) [reply packages/kilt-did/src/types.ts:167]: will do. I'm also not sure what the role of the DidRecord currently is as we have the DidDetails now

[the remark this answers is not in the document]
Comment (rflechtner) [reply packages/kilt-did/src/types.ts]: done

**Comment by Diiaablo95** [comment]
@tjwelde the following tests are still failing, should we merge anyways?

- unit test for Node 14.x fail because of not enough coverage
- unit test for Node 16.x fail for some reason, do we want to address the issues here?
- integration tests with latest mashnet release fail because of breaking changes
- examples and getting started fail, do we want to address the issues here?

If the PR is fine as it is, then I would say we can merge.

**Comment by tjwelde** [comment]
Yes, let's handle these issues after integrating the off-chain DID

**Review by tjwelde — APPROVED** [review]
LGTM

**Pull request merged**